test/e2e/coderef/code_reference_test.go

```diff
@@ -0,0 +1,284 @@
+	}
+	_, err := client.UpdateCodeReference(context.Background(), updateReq)
+	assert.NoError(t, err)
+
+	// Get and verify update
+	getResp := getCodeReference(t, client, &coderefproto.GetCodeReferenceRequest{
+		Id:            createResp.CodeReference.Id,
+		EnvironmentId: *environmentID,
+	})
+	assert.Equal(t, updateReq.FilePath, getResp.CodeReference.FilePath)
+	assert.Equal(t, updateReq.LineNumber, getResp.CodeReference.LineNumber)
+	assert.Equal(t, updateReq.CodeSnippet, getResp.CodeReference.CodeSnippet)
+	assert.Equal(t, updateReq.ContentHash, getResp.CodeReference.ContentHash)
+	assert.Equal(t, updateReq.RepositoryName, getResp.CodeReference.RepositoryName)
+	assert.Equal(t, updateReq.RepositoryOwner, getResp.CodeReference.RepositoryOwner)
+	assert.Equal(t, updateReq.RepositoryType, getResp.CodeReference.RepositoryType)
+	assert.Equal(t, updateReq.RepositoryBranch, getResp.CodeReference.RepositoryBranch)
+	assert.Equal(t, updateReq.CommitHash, getResp.CodeReference.CommitHash)
+}
+
+func TestListCodeReferences(t *testing.T) {
+	t.Parallel()
+	client := newCodeRefClient(t)
+	featureClient := newFeatureClient(t)
+
+	// First create a feature
+	featureID := createFeatureID(t)
+	createFeature(t, featureClient, featureID)
+
+	// Create multiple code references
+	createReq1 := newCreateCodeReferenceRequest(featureID)
+	createCodeReference(t, client, createReq1)
+	time.Sleep(time.Second)
+	createReq2 := newCreateCodeReferenceRequest(featureID)
+	createCodeReference(t, client, createReq2)
+
+	// List code references
+	listReq := &coderefproto.ListCodeReferencesRequest{
+		FeatureId:      featureID,
+		EnvironmentId:  *environmentID,
+		PageSize:       10,
+		Cursor:         "0",
+		OrderBy:        coderefproto.ListCodeReferencesRequest_CREATED_AT,
+		OrderDirection: coderefproto.ListCodeReferencesRequest_DESC,
+	}
+	listResp, err := client.ListCodeReferences(context.Background(), listReq)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(listResp.CodeReferences))
+}
+
+func TestDeleteCodeReference(t *testing.T) {
+	t.Parallel()
+	client := newCodeRefClient(t)
+	featureClient := newFeatureClient(t)
+
+	// First create a feature
+	featureID := createFeatureID(t)
+	createFeature(t, featureClient, featureID)
+
+	// Create code reference
+	createReq := newCreateCodeReferenceRequest(featureID)
+	createResp := createCodeReference(t, client, createReq)
+
+	// Delete code reference
+	deleteReq := &coderefproto.DeleteCodeReferenceRequest{
+		Id:            createResp.CodeReference.Id,
+		EnvironmentId: *environmentID,
+	}
+	_, err := client.DeleteCodeReference(context.Background(), deleteReq)
+	assert.NoError(t, err)
+
+	// Verify deletion
+	getReq := &coderefproto.GetCodeReferenceRequest{
+		Id:            createResp.CodeReference.Id,
+		EnvironmentId: *environmentID,
+	}
+	_, err = client.GetCodeReference(context.Background(), getReq)
+	assert.Error(t, err) // Should return error as code reference is deleted
```

Comment:
We must check the response code in case the backend returns any other error than not found.